### PR TITLE
Switch to Stack registry V2 API and add support the new stack YAML metadata fields

### PR DESCRIPTION
**Pull request by kke**
Makes the CLI communicate to the stack registry via the new V2 API which offers some benefits, such as speed improvement, better searches, JSONAPI compliance, gzip response compression and private stacks for organizations.

Adds optional metadata fields to Stack YAML:

```yaml
stack: user/stack
version: 1.2.3
services:
  redis:
    image: redis:latest
meta: # (optional)
  description: Short description / summary # (optional)
  icon: A URL to an SVG or PNG image to be used as an icon # (optional).
  home: The URL of this project's home page # (optional)
  source: The URL of this project's source code # (optional)
  readme: The README.md for this stack file (URL or markdown) # (optional) 
  issues: The URL of this project's issue tracking # (optional)
  required_kontena_version: ">= 1.5.0" # (optional)
  tags: # (optional)
    - A list of tags 
  app_version: # (optional)
    mongo: "3.2" 
    app: 1.2.3
  maintainers: # (optional)
    - name: name 
      email: 
      url: 
```

If the stack metadata has `required_kontena_version` defined, the `kontena stack install` and `kontena stack upgrade` commands check master version before the operation and prompt for confirmation unless `--force` is given.

### CODE REVIEW

File: cli/lib/kontena/cli/stacks/registry/create_command.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+require_relative '../common'
+
+module Kontena::Cli::Stacks::Registry
+  class CreateCommand < Kontena::Command
+    include Kontena::Cli::Common
+    include Kontena::Cli::Stacks::Common
+    include Kontena::Cli::Stacks::Common::RegistryNameParam
+
+    banner "Changes stack settings on registry"
```
Comment (jakolehm) [inline cli/lib/kontena/cli/stacks/registry/create_command.rb]: Misleading banner?

File: cli/lib/kontena/cli/stacks/registry/update_command.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+module Kontena::Cli::Stacks::Registry
+  class UpdateCommand < Kontena::Command
+    include Kontena::Cli::Common
+    include Kontena::Cli::Stacks::Common
+    include Kontena::Cli::Stacks::Common::RegistryNameParam
+
+    banner "Changes stack settings on registry"
+
+    option '--make-private', :flag, "Change visibility to private"
```
Comment (jakolehm) [inline cli/lib/kontena/cli/stacks/registry/update_command.rb]: Nope. `make-something` is clear indicator that this should be included in command name.

File: cli/lib/kontena/cli/stacks/stack_name.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ module Kontena::Cli::Stacks
   class StackName
     # A class for parsing stack name strings, such as kontena/foo:1.0.0
 
-    attr_reader :user, :stack, :version
+    attr_accessor :user, :stack, :version
```
Comment (jakolehm) [inline cli/lib/kontena/cli/stacks/stack_name.rb]: Why?

File: cli/lib/kontena/debug_instrumentor.rb [hunk 4]
```diff
@@ -39,12 +41,19 @@ def self.instrument(name, params = {}, &block)
       end
 
       if params[:body] && !params[:body].empty?
+        if params[:headers]['Content-Encoding'].to_s =~ /gzip/
+          body_content = Zlib::GzipReader.new(StringIO.new(params[:body])).read
```
Comment (jakolehm) [inline cli/lib/kontena/debug_instrumentor.rb]: Why is this here? Probably should be a separate PR?

File: cli/lib/kontena/cli/stacks/registry/update_command.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    option '--make-public', :flag, "Change visibility to ppublic"
```
Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/update_command.rb]: Also this seems to say ppublic

File: cli/lib/kontena/cli/stacks/stack_name.rb (hunk 3 above)
Comment (kke) [reply cli/lib/kontena/cli/stacks/stack_name.rb]: So that it's possible to `x = StackName.new('user/stack'); x.version = '1.2.3'`.

**Comment by kke** [comment]
The e2e will not pass before stack reg v2 is deployed to production.

**Comment by kke** [comment]
Test steps:

## V1 client compatibility

- [ ] Possible to create versions to self owned public stacks
- [ ] Possible to show public v2 stacks that have meta
- [ ] Possible to pull self-owned private stacks
- [ ] Possible to search public stacks
- [ ] Possible to search self owned private stacks
- [ ] Possible to pull public stacks using org/stackname
- [ ] Possible to pull public stacks using org/stackname:version
- [ ] Possible to pull self owned private stacks using org/stackname
- [ ] Possible to pull self owned private stacks using org/stackname:version
- [ ] Possible to delete self owned stack versions
- [ ] Not possible to delete non-self owned stack versions
- [ ] Not possible to pull deleted stack versions
- [ ] Possible to delete self owned stacks
- [ ] Not possible to delete non-self owned stacks
- [ ] Not possible to pull versions from a deleted stack
- [ ] Deleting a stack version does not delete other versions or stacks
- [ ] Deleting a stack does not delete other stacks
- [ ] Validation errors are intelligible when pushing an invalid stack yaml
- [ ] Errors are intelligible when pushing to non-owned stack
- [ ] Errors are intelligible when repushing an existing stack version

## V2 client

- [ ] Possible to create private stacks without versions
- [ ] Possible to create public stacks without versions
- [ ] Possible to create versions to self owned public stacks
- [ ] Possible to create versions to self owned private stacks
- [ ] Possible to create versions with meta
- [ ] Possible to show public stacks and their meta
- [ ] Possible to show self-owned private stacks and their meta
- [ ] Not possible to show non self-owned private stacks and their meta
- [ ] Possible to search public stacks
- [ ] Possible to search self owned private stacks
- [ ] Possible to search stacks using description
- [ ] Possible to search stacks using tags
- [ ] Possible to convert self owned public stacks to private
- [ ] Possible to convert self owned private stacks to public
- [ ] Possible to pull public stacks using org/stackname
- [ ] Possible to pull public stacks using org/stackname:version
- [ ] Possible to pull self owned private stacks using org/stackname
- [ ] Possible to pull self owned private stacks using org/stackname:version
- [ ] Possible to delete self owned stack versions
- [ ] Not possible to delete non-self owned stack versions
- [ ] Not possible to pull deleted stack versions
- [ ] Possible to delete self owned stacks
- [ ] Not possible to delete non-self owned stacks
- [ ] Not possible to pull versions from a deleted stack
- [ ] Deleting a stack version does not delete other versions or stacks
- [ ] Deleting a stack does not delete other stacks
- [ ] Validation errors are intelligible when pushing an invalid stack yaml
- [ ] Errors are intelligible when pushing to non-owned stack
- [ ] Errors are intelligible when repushing an existing stack version

**Comment by kke** [comment]
Bonus question: What to do with username owned stacks? Public still allowed?

**Comment by kke** [comment]
Test plan completed, some errors found and fixed.

File: cli/lib/kontena/stacks_cache.rb [hunk 6]
```diff
@@ -62,39 +57,39 @@ class RegistryClientFactory
...
         else
-          dputs "Retrieving #{stack.stack}:#{stack.version} from registry"
-          content = client.pull(stack.stack, stack.version)
-          yaml    = ::YAML.safe_load(content, [], [], true, stack.stack)
-          new_stack = CachedStack.new(yaml['stack'], yaml['version'])
+          dputs "Retrieving #{stack.stack_name} from registry"
+          content = client.pull(stack_name)
+          yaml    = ::YAML.safe_load(content, [], [], true, stack.stack_name.to_s)
+          new_stack_name = Kontena::Cli::Stacks::StackName.new(yaml['stack'])
+          new_stack_name.version = yaml['version']
```
Comment (jakolehm) [inline cli/lib/kontena/stacks_cache.rb]: I don't quite understand reasons behind this. If `StackName` needs `version` then why it's not passed in `.new`?

Comment (kke) [reply cli/lib/kontena/stacks_cache.rb]: Then this would have to be:

```
new_stack_name = Kontena::Cli::Stacks::StackName.new([yaml['stack'], yaml['version']].compact.join(':'))
```

or the `StackName#initialize` would need to accept an optional version argument.

Comment (kke) [reply cli/lib/kontena/stacks_cache.rb]: Hmm actually it does.

File: cli/lib/kontena/cli/stacks/stack_name.rb (hunk 3 above)
Comment (jakolehm) [reply cli/lib/kontena/cli/stacks/stack_name.rb]: I don't see that setter is needed because version can be passed via `.new`?

**Comment by kke** [comment]
E2e went green

**Comment by kke** [comment]
Doc issue: https://github.com/kontena/docs/issues/90

**Comment by kke** [comment]
Added some e2e, there's already some e2e's that install stacks from the registry.

All create/update/delete actions would need a cloud token and as the registry does not accept duplicate versions, it would have to bump the stack version number or do some other magic.

**Comment by kke** [comment]
One thing to think about is that this leaks some cloud fluff to the core CLI.

**Comment by kke** [comment]
image

**Comment by kke** [comment]
Failing e2e unrelated
